### ISSUE TITLE
Device file input

### ISSUE THREAD

**Issue by Lakritzator** [masmu/pulseaudio-dlna#176]
I know the name of the project is **pulseaudio**-dlna...
But I was wondering if it would be possible to get the same functionality without the pulseaudio, by creating device files where e.g. MPD could write to.

The main reason for this is that I would like to have the functionality of pulseaudio-dlna on a headless ubuntu server with MPD and I am having a lot of problems (read: didn't manage to) making it work. Somehow pulseaudio/d-bus does seem like a lot of overhead for streaming audio content which is most likely already in the right format.

Although I am a capable C# and Java developer, I never really got Python...

Any chance of making this happen? Doesn't have to be in the same repo...
My main focus is Ubuntu headless running MPD streaming to one or multiple Chromecasts (MPD has output selection)

**Comment by masmu** [comment]
Could you get me a small bash example which explains your idea a little bit more?

**Comment by Lakritzator** [comment]
I don't know how I should explain this in bash, but what I mean is that when start pulseaudio-dlna it will create fifo (First In First Out) files for every found DLNA / Chromecast (etc) in /dev
I can than configure MPD to have output(s) which write raw PCM to this/these file(s), and here your application would pick this up and send it to the DLNA or Chromecast.

Does this make it more clear?

**Comment by masmu** [comment]
Is _mpd_ capable of writing to _fifos_? And if yes, which format would it write? Are there some you can choose of?

**Comment by Lakritzator** [comment]
Yes, MPD is capable of writing to FIFOs, you can configure an output and specify the format like this:
```
audio_output {
  type   "fifo"
  name   "FIFO"
  path   "/tmp/mpd.fifo"
  format "44100:16:2"
}
```
The format is "raw" PCM, as far as I know there is no possibility of having a diffent codec.
Some details on specifying the output are here: http://www.musicpd.org/doc/user/config_audio_outputs.html

The possible plugins (type) you can use are documented here: http://www.musicpd.org/doc/user/output_plugins.html

Maybe "pipe" is also a possibility, this will start a command and pipe the audio to it.

Anyway, it would be really nice to have a way to output audio without having a dependency on pulseaudio. This would also make it possible to use mopidy or other applications on a NAS.

**Comment by masmu** [comment]
Well, I guess this is far from that projects scope. I thought about what code need to be changed to realize your idea and it is quite a lot. And all those changes would not fit into that projects direction.

**Possibility 1**
Or you could install an X server and leave a logged in user to ensure that a pulseaudio process is running in _session mode_. By far the easiest solution.

**Possiblity 2**
What you can do is: Write your own little HTTP server which serves the fifo's data when requesting a specific URL. It should:
  - set the appropriate DLNA headers
  - start an process (e.g.: `cat /dev/ff_mpd | lame -b 192 -r -`)
  - serve its data

With the following script you could instruct your devices to play that url: (Taken from [feature/google-cast-groups/scripts/radio.py](https://github.com/masmu/pulseaudio-dlna/blob/feature/google-cast-groups/scripts/radio.py))
```python
import pulseaudio_dlna
import pulseaudio_dlna.holder
import pulseaudio_dlna.plugins.upnp
import pulseaudio_dlna.plugins.chromecast
import pulseaudio_dlna.codecs

PLUGINS = [
    pulseaudio_dlna.plugins.upnp.DLNAPlugin(),
    pulseaudio_dlna.plugins.chromecast.ChromecastPlugin(),
]

holder = pulseaudio_dlna.holder.Holder(self.PLUGINS)
holder.search(ttl=5)
for device in holder.devices.values():
    device.play(url, pulseaudio_dlna.codecs.Mp3Codec())
```

**Possibility 3**
Get pulseaudio running in _system mode_. Someone how tried that summed up his experiences [here](https://github.com/masmu/pulseaudio-dlna/issues/159) when I remember correctly.

**Comment by Lakritzator** [comment]
Thanks for investicating this for me, apreciate it. I understand if it doesn't fit in the scope.

About the possibilities:
1) I might look into that, but like I said I find the need to have both a X server and Pulse-Audio as "dependency" just to send audio to a client :smiley:
2) I might try to write something for Core .NET in the near future, I do need some kind of project to play around with it a bit.
3) I tried this, and got stuck somehow in a state where pulseaudio-dlna didn't want to start. I went from problems with the DISPLAY variable to the next one with the DBUS address thingy...

Best wishes,
Robin
